raiden/network/transport/matrix.py

```diff
@@ -1061,22 +1095,68 @@ def _get_user(self, user: Union[User, str]) -> User:
             user = self._client.get_user(user)
         return user
 
-    def _set_room_id_for_address(self, address: Address, room_id: Optional[str]):
-        """ Uses GMatrixClient.set_account_data to keep updated mapping of addresses->rooms """
-        address_hex = to_checksum_address(address)
-        _address_to_room_id = self._client.account_data.get('network.raiden.rooms', {})
-        if room_id != _address_to_room_id.get(address_hex):
-            if room_id:
-                _address_to_room_id[address_hex] = room_id
-            else:
-                _address_to_room_id.pop(address_hex, None)
-            self._client.set_account_data('network.raiden.rooms', _address_to_room_id)
-
-    def _get_room_id_for_address(self, address: Address) -> Optional[str]:
-        """ Uses GMatrixClient.get_account_data to get updated mapping of addresses->rooms """
-        address_hex = to_checksum_address(address)
-        room_id = self._client.account_data.get('network.raiden.rooms', {}).get(address_hex)
-        if room_id and room_id not in self._client.rooms:
-            self._set_room_id_for_address(address, None)
-            return None
-        return room_id
+    def _set_room_id_for_address(self, address: Address, room_id: Optional[_RoomID]):
+        """ Uses GMatrixClient.set_account_data to keep updated mapping of addresses->rooms
+
+        If room_id is falsy, clean list of rooms. Else, push room_id to front of the list """
+
+        assert not room_id or room_id in self._client.rooms, 'Invalid room_id'
+        address_hex: AddressHex = to_checksum_address(address)
+
+        # no need to deepcopy, we don't modify lists in-place
+        _address_to_room_ids: Dict[AddressHex, List[_RoomID]] = self._client.account_data.get(
+            'network.raiden.rooms',
+            {},
+        ).copy()
+
+        if not room_id:  # falsy room_id => clear list
+            _address_to_room_ids.pop(address_hex, None)
+        else:
+            # filter_private=False to preserve public rooms on the list, even if we require privacy
+            room_ids = self._get_room_ids_for_address(address, filter_private=False)
+            # push to front
+            _address_to_room_ids[address_hex] = [room_id] + [r for r in room_ids if r != room_id]
+
+        # if changed:
+        if self._client.account_data.get('network.raiden.rooms', {}) != _address_to_room_ids:
+            self._client.set_account_data('network.raiden.rooms', _address_to_room_ids)
```

Comment:
No need for a lock, because there's no IO/context-switch between `get`ting the dict data (which is stored locally inside the client) and setting it (`_client.set_account_data` sets the internal dict before sending an update http request)

Reply:
I'm not so sure about that. `to_checksum_address` uses keccak which is provided by a C extension, that could very well be a context switching point for gevent.

Reply:
I don't think so, but will move `_get_room_ids_for_address` call to before `_address_to_room_ids` definition, to ensure no context happens between it and `_client.set_account_data`